In [1]:
import pandas as pd
import pickle
import preprocess_utils

from_database = False

if from_database:
    # Load tables from database 
    dfs = preprocess_utils.load_all_tables()
else:
    # Open the file in binary read mode and load the pickle data
    with open('preprocessed_data/dfs.pickle', 'rb') as f:
        dfs = pickle.load(f)
        for key, df in dfs.items():
            print(f"DataFrame {key:<20} has {len(df):<6} rows and {len(df.columns):<6} columns.")

DataFrame abuse_report_types   has 4      rows and 4      columns.
DataFrame abuse_reports        has 0      rows and 8      columns.
DataFrame ar_internal_metadata has 1      rows and 4      columns.
DataFrame articles             has 400    rows and 33     columns.
DataFrame assertion_types      has 10     rows and 11     columns.
DataFrame assertion_versions   has 19668  rows and 10     columns.
DataFrame assertions           has 13941  rows and 14     columns.
DataFrame assertions_genes     has 2053   rows and 2      columns.
DataFrame assertions_tags      has 403    rows and 2      columns.
DataFrame assertions_technique_types has 0      rows and 2      columns.
DataFrame assertions_techniques has 0      rows and 2      columns.
DataFrame assessment_types     has 13     rows and 9      columns.
DataFrame assessments          has 0      rows and 6      columns.
DataFrame authors              has 797    rows and 10     columns.
DataFrame banned_orcid_users   has 0      rows and 5   

In [2]:
if from_database:
    # Save tables to pickle
    with open('preprocessed_data/dfs.pickle', 'wb') as f:
        pickle.dump(dfs, f)

    # load it back with:
    with open('preprocessed_data/dfs.pickle', 'rb') as f:
        dfs = pickle.load(f)


In [3]:
def clean_df(df):
    # Create a copy of the DataFrame at the start
    df = df.copy()
    
    columns_to_remove = ['user_id', 'orcid_user_id', 
                    'created_at', 'updated_at', 'assertion_updated_at', 
                    'workspace_id', 'user_id', 'doi', 'organism_id', # 'pmid', 
                    'all_tags_json', 'obsolete', 'ext', 'badge_classes','pluralize_title',
                    'can_attach_file', 'refresh_side_panel', 'icon_classes', 'btn_classes']
    patterns_to_remove = ['validated', 'filename', 'obsolete_article']
    
    # Remove existing columns
    existing_cols = [col for col in columns_to_remove if col in df.columns]
    if existing_cols:
        df = df.drop(existing_cols, axis=1)
    
    # Remove pattern-matched columns
    pattern_cols = []
    for pattern in patterns_to_remove:
        pattern_cols.extend([c for c in df.columns if pattern in c])
    if pattern_cols:
        df = df.drop(pattern_cols, axis=1)
    
    return df

def truncate_string(s, max_length=20):
    """Truncate string to max_length characters."""
    if isinstance(s, str) and len(s) > max_length:
        return s[:max_length] + '...'
    return s

 ## 2. Process journal articles from the database

In [4]:
clean_df(dfs["articles"])

id                                        authors_txt  \
0    3347                                  Rizki RM;Rizki TM   
1    1987             Leulier F;Lhocine N;Lemaitre B;Meier P   
2    2761  Kim YS;Han SJ;Ryu JH;Choi KH;Hong YS;Chung YH;...   
3    2649                      Tingvall TO;Roos E;Engström Y   
4    3441                   Bhaskar V;Valentine SA;Courey AJ   
..    ...                                                ...   
395  1882        Derré I;Pypaert M;Dautry-Varsat A;Agaisse H   
396  1815    Schmidt RL;Trejo TR;Plummer TB;Platt JL;Tang AH   
397  1884     Avet-Rochex A;Perrin J;Bergeret E;Fauvarque MO   
398  2041                Tanji T;Ohashi-Kobayashi A;Natori S   
399  1926  Kurucz E;Márkus R;Zsámboki J;Folkl-Medzihradsz...   

                                                 title  journal_id      pmid  \
0    Cell interactions in the differentiation of a ...         517    111992   
1    The Drosophila inhibitor of apoptosis protein ...         200  16894030   
2    Lipopolysaccharide-activated kinase, an essent...          99  10636911   
3    The GATA factor Serpent is required for the on...          98  11274409   
4    A functional interaction between dorsal and co...          99  10660560   
..                                                 ...         ...       ...   
395  RNAi screen in Drosophila cells reveals the in...         121  17967059   
396  Infection-induced proteolysis of PGRP-LC contr...         129  18308747   
397  Rac2 is a major actor of Drosophila resistance...         186  17903178   
398  Participation of a galactose-specific C-type l...         428  16475980   
399  Nimrod, a putative phagocytosis receptor with ...         106  17363253   

    volume issue                                           abstract  year  \
0       12     3  The cellular events in the formation of melano...  1979   
1       26    21  The founding member of the inhibitor of apopto...  2006   
2      275     3  Eukaryotic organisms use a similar Rel/NF-kapp...  2000   
3       98     7  Innate immunity in Drosophila is characterized...  2001   
4      275     6  To identify proteins that regulate the functio...  2000   
..     ...   ...                                                ...   ...   
395      3    10  Chlamydia spp. are intracellular obligate bact...  2007   
396     22     3  The Drosophila immune deficiency (IMD) pathway...  2008   
397     12    10  Pathogen recognition and engulfment by phagocy...  2007   
398    396     1  A galactose-specific C-type lectin has been pu...  2006   
399     17     7  The hemocytes, the blood cells of Drosophila, ...  2007   

    published_at     key                                          affs_json  \
0            NaT  4yet3q                                            [[],[]]   
1            NaT  x8x1ej  [["The Breakthrough Toby Robins Breast Cancer ...   
2            NaT  3klzn3  [["Laboratory of Immunology, Medical Research ...   
3            NaT  vi2oc1  [["Department of Molecular Biology and Functio...   
4            NaT  wgssyx  [["Department of Chemistry and Biochemistry, U...   
..           ...     ...                                                ...   
395          NaT  pgclng  [["Section of Microbial Pathogenesis, Yale Uni...   
396          NaT  b8fmj9  [["Department of Biochemistry and Molecular Bi...   
397          NaT  gmmelt  [["Commissariat à l'Energie Atomique, DSV, iRT...   
398          NaT  vjzt83  [["Department of Cell Biochemistry, Graduate S...   
399          NaT  lvm3ua  [["Institute of Genetics, Biological Research ...   

    references_txt additional_context  num  large_scale  nber_tables  \
0             None               None  371        False          0.0   
1             None               None  145        False          0.0   
2             None               None  311        False          0.0   
3             None               None  292        False          1.0   
4             None               None  378 

In [5]:
# Process articles
articles = clean_df(dfs["articles"])
articles = articles.rename(columns={"id": "article_id"})
# add journals to articles:
journals = clean_df(dfs["journals"])
journals = journals.drop('tag', axis=1).rename(columns={"id": "journal_id", "name": "journal_name"})
articles = articles.merge(journals, on="journal_id", how="left", suffixes=('', '_journal')).drop("journal_id", axis=1)


# Extract first and last authors from authors_txt
def extract_authors(authors_txt):
    if isinstance(authors_txt, str) and authors_txt:
        authors_list = authors_txt.split(';')
        first_author = authors_list[0].strip() if authors_list else None
        last_author = authors_list[-1].strip() if authors_list else None
        return pd.Series([first_author, last_author])
    return pd.Series([None, None])

# Apply the function to create new columns
articles[['first_author_extracted', 'last_author_extracted']] = articles['authors_txt'].apply(extract_authors)

# Check the results
print(f"Total articles: {len(articles)}")
print(f"Articles with first author: {articles['first_author_extracted'].notna().sum()}")
print(f"Articles with last author: {articles['last_author_extracted'].notna().sum()}")

articles = articles.drop(columns=["key", "references_txt", "additional_context", "num",  "nber_panels", "large_scale", "nber_tables", "published_at"])


# Display the first few rows to verify
articles[['authors_txt', 'first_author_extracted', 'last_author_extracted']].head()

Total articles: 400
Articles with first author: 400
Articles with last author: 400


authors_txt first_author_extracted  \
0                                  Rizki RM;Rizki TM               Rizki RM   
1             Leulier F;Lhocine N;Lemaitre B;Meier P              Leulier F   
2  Kim YS;Han SJ;Ryu JH;Choi KH;Hong YS;Chung YH;...                 Kim YS   
3                      Tingvall TO;Roos E;Engström Y            Tingvall TO   
4                   Bhaskar V;Valentine SA;Courey AJ              Bhaskar V   

  last_author_extracted  
0              Rizki TM  
1               Meier P  
2                Lee WJ  
3            Engström Y  
4             Courey AJ

In [6]:
import json

# We need to analyze and clean up affiliations in the articles dataframe

def extract_primary_affiliation(affs_json_str):
    """Extract the first affiliation from the JSON string, safely handle bad JSON format."""
    try:
        # Parse the JSON string to a Python object
        affs_list = json.loads(affs_json_str)
        
        # Get the first non-empty affiliation
        for aff in affs_list:
            if aff and isinstance(aff, list) and len(aff) > 0 and aff[0].strip():
                return aff[0].strip()
        
        return None  # No valid affiliation found
    except (json.JSONDecodeError, TypeError, IndexError):
        # Handle poorly formatted JSON or other errors
        return None

# Create a new column with clean affiliations
articles['primary_affiliation'] = articles['affs_json'].apply(extract_primary_affiliation)

# Count articles with multiple vs. single affiliations
def count_affiliations(affs_json_str):
    """Count the number of non-empty affiliations in the JSON string."""
    try:
        affs_list = json.loads(affs_json_str)
        count = sum(1 for aff in affs_list if aff and isinstance(aff, list) and len(aff) > 0 and aff[0].strip())
        return count
    except (json.JSONDecodeError, TypeError, IndexError):
        return 0

articles['affiliation_count'] = articles['affs_json'].apply(count_affiliations)

# Identify articles with multiple affiliations
multiple_aff_articles = articles[articles['affiliation_count'] > 1]

# Print summary statistics
print(f"Total articles: {len(articles)}")
print(f"Articles with 1 affiliation: {len(articles[articles['affiliation_count'] == 1])}")
print(f"Articles with multiple affiliations: {len(multiple_aff_articles)}")
print(f"Articles with no valid affiliations: {len(articles[articles['affiliation_count'] == 0])}")

# Sample of articles with multiple affiliations
if not multiple_aff_articles.empty:
    print("\nSample of articles with multiple affiliations:")
    sample = multiple_aff_articles.head(5)
    for _, row in sample.iterrows():
        print(f"Article ID: {row['article_id']}, Title: {row['title'][:50]}...")
        try:
            affs = json.loads(row['affs_json'])
            for i, aff in enumerate(affs):
                if aff and isinstance(aff, list) and len(aff) > 0 and aff[0].strip():
                    print(f"  Affiliation {i+1}: {aff[0]}")
        except:
            print("  Error parsing affiliations")
        print()

Total articles: 400
Articles with 1 affiliation: 387
Articles with multiple affiliations: 0
Articles with no valid affiliations: 13


In [7]:
articles = articles.drop(columns=["affs_json", "affiliation_count"])

In [8]:
articles

article_id                                        authors_txt  \
0          3347                                  Rizki RM;Rizki TM   
1          1987             Leulier F;Lhocine N;Lemaitre B;Meier P   
2          2761  Kim YS;Han SJ;Ryu JH;Choi KH;Hong YS;Chung YH;...   
3          2649                      Tingvall TO;Roos E;Engström Y   
4          3441                   Bhaskar V;Valentine SA;Courey AJ   
..          ...                                                ...   
395        1882        Derré I;Pypaert M;Dautry-Varsat A;Agaisse H   
396        1815    Schmidt RL;Trejo TR;Plummer TB;Platt JL;Tang AH   
397        1884     Avet-Rochex A;Perrin J;Bergeret E;Fauvarque MO   
398        2041                Tanji T;Ohashi-Kobayashi A;Natori S   
399        1926  Kurucz E;Márkus R;Zsámboki J;Folkl-Medzihradsz...   

                                                 title      pmid volume issue  \
0    Cell interactions in the differentiation of a ...    111992     12     3   
1    The Drosophila inhibitor of apoptosis protein ...  16894030     26    21   
2    Lipopolysaccharide-activated kinase, an essent...  10636911    275     3   
3    The GATA factor Serpent is required for the on...  11274409     98     7   
4    A functional interaction between dorsal and co...  10660560    275     6   
..                                                 ...       ...    ...   ...   
395  RNAi screen in Drosophila cells reveals the in...  17967059      3    10   
396  Infection-induced proteolysis of PGRP-LC contr...  18308747     22     3   
397  Rac2 is a major actor of Drosophila resistance...  17903178     12    10   
398  Participation of a galactose-specific C-type l...  16475980    396     1   
399  Nimrod, a putative phagocytosis receptor with ...  17363253     17     7   

                                              abstract  year  \
0    The cellular events in the formation of melano...  1979   
1    The founding member of the inhibitor of apopto...  2006   
2    Eukaryotic organisms use a similar Rel/NF-kapp...  2000   
3    Innate immunity in Drosophila is characterized...  2001   
4    To identify proteins that regulate the functio...  2000   
..                                                 ...   ...   
395  Chlamydia spp. are intracellular obligate bact...  2007   
396  The Drosophila immune deficiency (IMD) pathway...  2008   
397  Pathogen recognition and engulfment by phagocy...  2007   
398  A galactose-specific C-type lectin has been pu...  2006   
399  The hemocytes, the blood cells of Drosophila, ...  2007   

                                          journal_name  impact_factor  \
0    Differentiation; research in biological diversity           3.39   
1                       Molecular and cellular biology           5.30   
2                  The Journal of biological chemistry           4.80   
3    Proceedings of the National Academy of Science...          11.10   
4                  The Journal of biological chemistry           4.80   
..                                                 ...            ...   
395                                     PLoS pathogens           6.70   
396  FASEB journal : official publication of the Fe...           4.80   
397  Genes to cells : devoted to molecular &amp; ce...           2.10   
398                            The Biochemical journal           4.10   
399                               Current biology : CB           9.20   

    first_author_extracted last_author_extracted  \
0                 Rizki RM              Rizki TM   
1                Leulier F               Meier P   
2                   Kim YS                Lee WJ   
3              Tingvall TO            Engström Y   
4                Bhaskar V             Courey AJ   
..                     ...                   ...   
395                Derré I             Agaisse H   
396             Schmidt RL               Tang AH   
397          Avet-Rochex A          Fauvarque MO   
398                Tanji 

In [9]:
articles.to_csv("preprocessed_data/articles_db.csv", index=False)

 ## 2. Process claims from the database

claims is merged with articles


In [10]:
# Main processing
claims = dfs["assertions"].copy()
print(len(claims))
claims = claims[claims["obsolete"] == False].copy()
print(len(claims))
claims = clean_df(claims)

id_cols = [col for col in claims.columns if "_id" in col]
print(id_cols)

claims = claims.merge(articles, on="article_id", how="left", suffixes=('', '_article'))
id_cols = [col for col in claims.columns if "_id" in col]
print(id_cols)


# Process assertion types
assertion_types = clean_df(dfs["assertion_types"])
assertion_types = assertion_types.rename(columns={"id": "assertion_type_id", "name": "assertion_type"})
claims = claims.merge(assertion_types, on="assertion_type_id", how="left", suffixes=('', '_assertion_type')).drop("assertion_type_id", axis=1)

# Process assessment types
assessment_types = clean_df(dfs["assessment_types"])
assessment_types = assessment_types.rename(columns={"id": "assessment_type_id", "name": "assessment_type"})
claims = claims.merge(assessment_types, on="assessment_type_id", how="left", suffixes=('', '_assessment_type')).drop("assessment_type_id", axis=1)

id_cols = [col for col in claims.columns if "_id" in col]
print(id_cols)

13941
13299
['article_id', 'assertion_type_id', 'assessment_type_id']


['article_id', 'assertion_type_id', 'assessment_type_id']
['article_id']


In [11]:
claims = claims.drop(['badge_tag_classes','description'], axis=1) # most not consistently used accross dataset
claims = claims.set_index('id')

In [12]:
# Update impact factor for "Proceedings. Biological sciences" and standardize the journal name
claims.loc[claims["journal_name"] == "Proceedings. Biological sciences", "impact_factor"] = 4.7 
claims.loc[claims["journal_name"] == "Proceedings. Biological sciences", "journal_name"] = "Proceedings Biological Sciences"

string_columns = claims.select_dtypes(include='object').columns
print(string_columns)
for col in string_columns:
    claims[col] = claims[col].apply(lambda x: x.replace('&amp;', '&') if isinstance(x, str) else x)

Index(['content', 'pmid', 'authors_txt', 'title', 'volume', 'issue',
       'abstract', 'journal_name', 'first_author_extracted',
       'last_author_extracted', 'primary_affiliation', 'assertion_type',
       'label', 'assessment_type'],
      dtype='object')


In [13]:

claims.to_csv('preprocessed_data/claims_db.csv')


string_columns = string_columns.drop(["assessment_type", "journal_name"])
df_truncated = claims.copy()
for col in string_columns:
    if col in df_truncated.columns:
        df_truncated[col] = df_truncated[col].apply(lambda x: truncate_string(x))

# Save truncated dataframe
df_truncated.to_csv('preprocessed_data/claims_db_truncated_for_llm.csv', index=False)
df_truncated

content  article_id  rank  pmid  \
id                                                       
5838   <p>Belozerov VE, Lin...      2049.0     1  None   
693                                 2517.0     1  None   
695                                 2517.0     1  None   
5840   Verified by many sub...      2204.0     1  None   
5839   Activation of a cell...      2204.0     1  None   
...                        ...         ...   ...   ...   
14468  The DIF antiserum us...      2898.0     1  None   
11330  Although independent...      1577.0     1  None   
11329  Spatzle (#gene:FBgn0...      1577.0     1  None   
11350  Although independent...      1577.0     1  None   
11349  Spatzle (#gene:FBgn0...      1577.0     3  None   

                   authors_txt                    title  pmid_article volume  \
id                                                                             
5838   Zhuang ZH;Zhou Y;Yu ...  Regulation of Drosop...    16014325.0     18   
693    Choe KM;Werner T;Stö...  Requirement for a pe...    11872802.0    296   
695    Choe KM;Werner T;Stö...  Requirement for a pe...    11872802.0    296   
5840   Zettervall CJ;Anderl...  A directed screen fo...    15381778.0    101   
5839   Zettervall CJ;Anderl...  A directed screen fo...    15381778.0    101   
...                        ...                      ...           ...    ...   
14468  Williams MJ;Rodrigue...  The 18-wheeler mutat...     9321392.0     16   
11330  Shia AK;Glittenberg ...  Toll-dependent antim...    19934223.0    122   
11329  Shia AK;Glittenberg ...  Toll-dependent antim...    19934223.0    122   
11350  Shia AK;Glittenberg ...  Toll-dependent antim...    19934223.0    122   
11349  Shia AK;Glittenberg ...  Toll-dependent antim...    19934223.0    122   

       issue                 abstract  ...  \
id                                     ...   
5838       4  The p38 mitogen-acti...  ...   
693     5566  Components of microb...  ...   
695     5566  Components of microb...  ...   
5840      39  An attack by a paras...  ...   
5839      39  An attack by a paras...  ...   
...      ...                      ...  ...   
14468     20  Mammals and insects ...  ...   
11330  Pt 24  In Drosophila, the h...  ...   
11329  Pt 24  In Drosophila, the h...  ...   
11350  Pt 24  In Drosophila, the h...  ...   
11349  Pt 24  In Drosophila, the h...  ...   

                                            journal_name impact_factor  \
id                                                                       
5838                                 Cellular signalling           4.8   
693                             Science (New York, N.Y.)          56.9   
695                             Science (New York, N.Y.)          56.9   
5840   Proceedings of the National Academy of Science...          11.1   
5839   Proceedings of the National Academy of Science...          11.1   
...                                                  ...           ...   
14468                                   The EMBO journal          11.4   
11330                            Journal of cell science           4.0   
11329                            Journal of cell science           4.0   
11350                            Journal of cell science           4.0   
11349                            Journal of cell science           4.0   

       first_author_extracted last_author_extracted      primary_affiliation  \
id                                                                             
5838                Zhuang ZH                 Ge BX  Signal Transduction ...   
693                   Choe KM           Anderson KV  Molecular Biology Pr...   
695                   Choe KM           Anderson KV  Molecular Biology Pr...   
5840            Zettervall CJ            Hultmark D  Umeå Centre for Mole...   
5839            Zettervall CJ            Hultmark D  Umeå Centre for Mole...   
...                       ...                   ...                      ...   
14468             William

## 3. Process authors from database

there are about two author per article... though the total of authors is 797 so some are missing.

In [14]:
a = clean_df(dfs["authors"])  
a[a["article_id"] == 2053]

id  article_id        name  sex  career_stage_id  leading_author  \
0   2        2053      Brun S    1                3           False   
1   3        2053  Lemaitre B    1                4            True   

   first_author  expertise_level_id  
0          True                   2  
1         False                   2

In [15]:
len(a)

797

In [16]:
a[a["name"] == "Lemaitre B"].head()

id  article_id        name  sex  career_stage_id  leading_author  \
1    3        2053  Lemaitre B    1                4            True   
7    8        3437  Lemaitre B    1                4            True   
14  16        2801  Lemaitre B    1                4            True   
30  36        2617  Lemaitre B    1                4            True   
44  56        2869  Lemaitre B    1                4            True   

    first_author  expertise_level_id  
1          False                   2  
7          False                   2  
14         False                   2  
30         False                   2  
44         False                   2

In [17]:
a[a["article_id"] == 2451]

id  article_id           name  sex  career_stage_id  leading_author  \
17    19        2451  De Gregorio E    1                3           False   
95    20        2451     Lemaitre B    1                5           False   
789  795        2451        Brey PT    1                5            True   

     first_author  expertise_level_id  
17           True                   2  
95          False                   2  
789         False                   2

In [18]:
authors = clean_df(dfs["authors"].copy())
authors
career_stage = dfs["career_stages"].rename(columns={"id": "career_stage_id"})
authors = authors.merge(career_stage, on="career_stage_id", how="left", suffixes=('', '_career_stage')).drop("career_stage_id", axis=1)
expertise_level = dfs["expertise_levels"].rename(columns={"id": "expertise_level_id"})
authors = authors.merge(expertise_level, on="expertise_level_id", how="left", suffixes=('', '_expertise_level')).drop("expertise_level_id", axis=1)
authors = authors.rename(columns={"name_career_stage": "career_stage", "name_expertise_level": "expertise_level", "id": "author_id"})

# Convert sex values: 1 to "Male", 0 to "Female"
authors.loc[authors["sex"] == 1, "sex"] = "Male"
authors.loc[authors["sex"] == 0, "sex"] = "Female"
authors["author_key"] =  authors["name"].str.lower().str.normalize('NFKD').str.encode('ascii', errors='ignore').str.decode('utf-8')

authors

author_id  article_id               name     sex  leading_author  \
0            2        2053             Brun S    Male           False   
1            3        2053         Lemaitre B    Male            True   
2            1        1650         Charroux B    Male           False   
3           97        2149        Söderhäll K    Male            True   
4            5        2278        Silverman N    Male           False   
..         ...         ...                ...     ...             ...   
792        797        2067         Cerenius L    Male            True   
793        798        2122         Higgins DE    Male            True   
794        799        2184  Mengin-Lecreulx D    Male            True   
795        800        2130           Moore KJ  Female            True   
796        700        1990           Matova N  Female           False   

     first_author career_stage expertise_level         author_key  
0            True     Post-doc     Experienced             brun s  
1           False    Junior PI     Experienced         lemaitre b  
2            True     Post-doc     Experienced         charroux b  
3           False    Senior PI     Experienced        soderhall k  
4            True     Post-doc        Newcomer        silverman n  
..            ...          ...             ...                ...  
792         False    Senior PI     Experienced         cerenius l  
793         False    Senior PI        Newcomer         higgins de  
794         False    Senior PI     Experienced  mengin-lecreulx d  
795         False    Junior PI        Newcomer           moore kj  
796          True     Post-doc        Newcomer           matova n  

[797 rows x 9 columns]

In [19]:
# check authors thare not both leading and first authors, or that are neither
authors[~(authors["leading_author"] ^ authors["first_author"])]

author_id  article_id           name     sex  leading_author  \
95          20        2451     Lemaitre B    Male           False   
194        127        2113   Schneider DS    Male           False   
195        197        3440     Lemaitre B    Male           False   
196        169        2184     Lemaitre B    Male           False   
243        250        1947   Schneider DS    Male            True   
256        260        1763   Schneider DS    Male           False   
257        252        2730  Shahabuddin M    Male           False   
287        303        3458     Hultmark D    Male            True   
329        251        2730   Schneider DS    Male            True   
334        378        1702    Williams MJ    Male            True   
342        365        3447       Rizki TM    Male            True   
350        363        3448       Rizki TM    Male            True   
355        364        3292       Rizki TM    Male            True   
356        370        3450       Gateff E  Female            True   
368        362        3273       Rizki TM    Male            True   
371        371        3449       Gateff E  Female            True   
473        483        2764       Imler JL    Male            True   
565        497        2122     Perrimon N    Male           False   
658        647        3446       Nappi AJ    Male            True   
686        702        1762      Wegner KM    Male            True   
741        720        2067    Söderhäll K    Male           False   
742        706        2130   Ezekowitz RA    Male           False   
790        792        3273    Bellotti RA    Male           False   

     first_author career_stage expertise_level     author_key  
95          False    Senior PI     Experienced     lemaitre b  
194         False    Junior PI     Experienced   schneider ds  
195         False    Junior PI     Experienced     lemaitre b  
196         False    Junior PI     Experienced     lemaitre b  
243          True    Senior PI     Experienced   schneider ds  
256         False    Senior PI     Experienced   schneider ds  
257         False    Junior PI        Newcomer  shahabuddin m  
287          True    Senior PI     Experienced     hultmark d  
329          True    Junior PI        Newcomer   schneider ds  
334          True    Junior PI     Experienced    williams mj  
342          True    Senior PI     Experienced       rizki tm  
350          True    Senior PI     Experienced       rizki tm  
355          True    Senior PI     Experienced       rizki tm  
356          True    Senior PI     Experienced       gateff e  
368          True     Post-doc     Experienced       rizki tm  
371          True    Senior PI     Experienced       gateff e  
473          True    Junior PI        Newcomer       imler jl  
565         False    Senior PI     Experienced     perrimon n  
658          True    Junior PI     Experienced       nappi aj  
686          True     Post-doc        Newcomer      wegner km  
741         False    Senior PI     Experienced    soderhall k  
742         False    Senior PI        Newcomer   ezekowitz ra  
790         False       Master        Newcomer    bellotti ra

In [20]:
#authors = authors.merge(articles, on="article_id", how="left", suffixes=('', '_article'))
#authors.sort_values(by=["article_id", "author_id", ])
def clean_author_keys(df,  column):
    replacements = {
        'ando': 'ando i',
        'derre': 'derre i',
        'markus': 'markus r'
    }
    
    # Replace values only where they exactly match
    df.loc[df[column].isin(replacements.keys()), column] = \
        df.loc[df[column].isin(replacements.keys()), column].map(replacements)
    
    return df

authors = clean_author_keys(authors, "author_key")

In [21]:
authors.to_csv('preprocessed_data/authors_db.csv', index=False)

In [22]:
authors.sort_values(by=["author_key"])

author_id  article_id          name     sex  leading_author  \
724        745        1535  Abdelsadik A    Male           False   
485        496        2122     Agaisse H    Male           False   
487        499        1882     Agaisse H    Male            True   
15          17        2297     Agaisse H    Male           False   
104        108        1752    Aggarwal K  Female           False   
..         ...         ...           ...     ...             ...   
754        784        1491         Zhu S    Male            True   
740        714        1756         Zhu S    Male            True   
653        621        1748         Zhu S    Male            True   
588        588        2025     Zhuang ZH    Male           False   
51          68        2049     Zhuang ZH    Male           False   

     first_author career_stage expertise_level    author_key  
724          True          PhD     Experienced  abdelsadik a  
485          True     Post-doc     Experienced     agaisse h  
487         False    Junior PI     Experienced     agaisse h  
15           True     Post-doc     Experienced     agaisse h  
104          True          PhD     Experienced    aggarwal k  
..            ...          ...             ...           ...  
754         False    Junior PI        Newcomer         zhu s  
740         False    Junior PI        Newcomer         zhu s  
653         False    Senior PI        Newcomer         zhu s  
588          True          PhD        Newcomer     zhuang zh  
51           True          PhD        Newcomer     zhuang zh  

[797 rows x 9 columns]

## 4. Preprocess the author files

Last files version:
- **0: First** `input_data/2025-03-28/2025-03-27_dropbox_6_ReproSci/0_last version/2025_Last version_March 19th/0_First author cleaned_March 18th.xlsx`
  - Pour les statistiques sur ceux qui deviennent PI, Il ne faut pas prendre en compte ceux qui sont surlignés en bleu car ils sont des premiers auteurs qui sont déjà PI. Normalement ce fichier est bon.  
Avec un peu d’approximation, je trouve que sur 291 first authors: 107 become PI et il y  16  ?? (indéterminés)et 13 qui sont déjà PI en bleu :  ce qui fait 107/262=>40,8% (262=291-16-13)  
- **2: Both** `input_data/2025-03-28/2025-03-27_dropbox_6_ReproSci/0_last version/2025_Last version_March 19th/2_2025_March 9th_stats_author.xlsx`
  - obtenu depuis le site
  -  tu trouveras les last avec de nouveau critères  pour historical (tradition), continuty, first and last.  (pour junior et senior il faut voir par articles dans la database)  
  - Feuille 1 : Parmi les first author (1er feuille) :  
    - Zhenting Zhang et Zhaolin Zhang  sont fusionnés en Zhang Z  
    - Zhipeng Wang et Zhi Wang sont fusionné en Wang Z  
    - Hedengren M et Hedengren M et Hedengren-Olcott M ne sont pas fusionnés
  - Feuille 3: pour les last authors:
    - Il faut fusionner Schneider DS et Schneider D -> c'est les mêmes
    - Historical lab (tradition) or not : il faut pendre le colonne Historical lab after 1998 et comparer 1 (historical, trained in hustorical lab) et 0 (non historical) mais ne pas prendre en compte 2.
    - La colonne E first and Last : compte les chercheurs qui apparaissent comme first author  et last authors. Il faudrait comparer 1 (last author qui ont été first author dans une autre labo mais pas PI)  à  0 (last author withouth being first author)  sans prendre en compte les 2 (first author who were already PI when first author).
  
- **3:citations** `input_data/2025-03-28/2025-03-27_dropbox_6_ReproSci/0_last version/2025_Last version_March 19th/3_2025_March 9th_citation_counts.xlsm`
  -  il s’agit des citations : je ne l’ai pas travaillé.  
 - **4:first&last** `input_data/2025-03-28/2025-03-27_dropbox_6_ReproSci/0_last version/2025_Last version_March 19th/4_working first and last.xlsx`
    -   ils’agit d’une liste de 20 auteurs qui sont first and last and mais ne pas prendre en compte ceux qui sont PI.  

 It seems that
 - sex -> FH
 - PhD Post-doc -> FH
 - Become a Pi -> FH
 - current job -> FH
 - MD -> FH
 - Affiliation -> Both
 - Country -> Both
 - Ivy League -> Both

In [23]:
# read an xlx file in pandas
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import  preprocess_utils

stat_author_fn = "input_data/2025-03-28/2025-03-27_dropbox_6_ReproSci/0_last version/2025_Last version_March 19th/2_2025_March 9th_stats_author.xlsx"

claims_df = pd.read_csv('preprocessed_data/claims_db_truncated_for_llm.csv')
major_claims_df = claims_df[claims_df['assertion_type'] == 'major_claim']

# Read Author info, which contains all the pairs
paper_auth_pairs = pd.read_excel('input_data/2025-02-14_last_xlsx/1_Triage_Last author.xlsx', sheet_name='Tri sans les doublons')
# Drop all columns with 'Unnamed' in the name
paper_auth_pairs = paper_auth_pairs.drop(columns=[col for col in paper_auth_pairs.columns if 'Unnamed' in col]).drop(columns=['Source'])
# rename RIZKI MT	to Rizki TM in column last author
paper_auth_pairs['last author'] = paper_auth_pairs['last author'].replace({'Rizki MT': 'Rizki TM','RIZKI MT': 'Rizki TM' })

paper_auth_pairs.loc[paper_auth_pairs["MD"] == "??", "MD"] = np.nan
paper_auth_pairs.loc[paper_auth_pairs["MD"] == "?", "MD"] = np.nan
paper_auth_pairs.loc[paper_auth_pairs["MD"] == 'O', "MD"] = False
paper_auth_pairs.loc[paper_auth_pairs["MD"] == 0, "MD"] = False
paper_auth_pairs.loc[paper_auth_pairs["MD"] == 1, "MD"] = True

paper_auth_pairs.loc[paper_auth_pairs["Become a Pi"] == "??", "Become a Pi"] = np.nan
paper_auth_pairs.loc[paper_auth_pairs["Become a Pi"] == 0, "Become a Pi"] = False
paper_auth_pairs.loc[paper_auth_pairs["Become a Pi"] == 1, "Become a Pi"] = True

paper_auth_pairs.loc[paper_auth_pairs["Ivy league"] == 0, "Ivy league"] = False
paper_auth_pairs.loc[paper_auth_pairs["Ivy league"] == 1, "Ivy league"] = True

paper_auth_pairs.loc[paper_auth_pairs["Sex"] == "??", "Sex"] = np.nan
paper_auth_pairs.loc[paper_auth_pairs["Sex"] == 1, "Sex"] = "Male"
paper_auth_pairs.loc[paper_auth_pairs["Sex"] == 0, "Sex"] = "Female"

paper_auth_pairs.loc[paper_auth_pairs["current job"] == "pi", "current job"] = "PI"
paper_auth_pairs.loc[paper_auth_pairs["current job"] == "Pi", "current job"] = "PI"
paper_auth_pairs.loc[paper_auth_pairs["current job"] == "?", "current job"] = np.nan
paper_auth_pairs.loc[paper_auth_pairs["current job"] == "??", "current job"] = np.nan

# create merge columns: lowercased and stripped of accents
paper_auth_pairs['first_author_key'] = paper_auth_pairs['first author'].str.lower()
paper_auth_pairs['first_author_key'] = paper_auth_pairs['first_author_key'].str.normalize('NFKD').str.encode('ascii', errors='ignore').str.decode('utf-8')
paper_auth_pairs = clean_author_keys(paper_auth_pairs, "first_author_key")


paper_auth_pairs.loc[:, 'last_author_key'] = paper_auth_pairs['last author'].str.lower()
paper_auth_pairs.loc[:, 'last_author_key'] = (paper_auth_pairs['last_author_key']
    .str.normalize('NFKD')
    .str.encode('ascii', errors='ignore')
    .str.decode('utf-8'))
#paper_auth_pairs.loc[:, 'last_author_key'] = paper_auth_pairs['last_author_key'].str.replace('rizki mt', 'rizki tm')
paper_auth_pairs = clean_author_keys(paper_auth_pairs, "last_author_key")


print(paper_auth_pairs["Sex"].unique(), paper_auth_pairs["MD"].unique(), paper_auth_pairs["Become a Pi"].unique(), paper_auth_pairs["Ivy league"].unique())


paper_auth_pairs.to_csv('input_data/2025-02-14_last_xlsx/1_Triage_Last author.csv', index=False)

first_authors_claims = pd.read_excel(stat_author_fn, sheet_name='First')
leading_authors_claims = pd.read_excel(stat_author_fn, sheet_name='Leading')
leading_authors_claims["Authorship"]= "Leading"
first_authors_claims["Authorship"]= "First"

authors_claims = pd.concat([leading_authors_claims, first_authors_claims])
authors_claims['Sex'] = authors_claims['Sex'].map({1: 'Male', 0: 'Female'})
authors_claims = authors_claims.drop(columns=[col for col in authors_claims.columns if '%' in col])
authors_claims = authors_claims.drop(columns=[col for col in authors_claims.columns if 'Unnamed' in col])

authors_claims = authors_claims.rename(columns={'Conituinity': 'Continuity', 
                                                "Partially verified":"Partially Verified", 
                                                "*Historical lab after 1998":"Historical lab after 1998"
                                                })

authors_claims['Historical lab'] = authors_claims['Historical lab'].astype('boolean')
authors_claims['Continuity'] = authors_claims['Continuity'].astype('boolean')
authors_claims = authors_claims.dropna(subset=['Name'])
# drop  Schneider DS+D from names as  Schneider DS and  Schneider D are also there in separate rows
authors_claims = authors_claims[~(authors_claims['Name'] == 'Schneider DS+D')]



authors_claims.to_csv('input_data/2025-02-14_last_xlsx/stats_author.csv', index=False)

['Female' 'Male' nan] [False nan True] [True False nan] [True False]


In [24]:
paper_auth_pairs["current job"].value_counts()

current job
PI                                100
Industry                           54
Academia                           24
Admin                              19
Senior Staff                       11
Retired                             8
Not in academia                     7
Deceased                            5
High school teacher                 5
Medical doctor                      4
Lecturer at university              4
Tech                                4
Science communication               3
Admin Academia                      2
Patent Lawyer                       2
Facility leader                     2
Museum                              2
School teacher                      2
Clinical research                   2
University Lecturer                 1
Science editor                      1
Clinical Researcher                 1
Adjunct Assistant Professor         1
nurse                               1
Hospital                            1
veterinary docto                    1


### A. Last authors

In [25]:
paper_auth_pairs

last author first author     Sex PhD Post-doc Become a Pi  \
0      Agaisse H        Derré  Female     Post-doc        True   
1    Aguilera RJ     Seong CS    Male          PhD       False   
2       Aigaki T      Tsuda M    Male          PhD        True   
3           Ando       Markus    Male          PhD       False   
4           Ando        Rus F  Female          PhD       False   
..           ...          ...     ...          ...         ...   
315        Yu XQ         Ao J     NaN          PhD         NaN   
316        Zhu S       Yuan Y    Male          PhD       False   
317        Zhu S       Tian C  Female          PhD        True   
318        Zhu S      Zhang Z  Female          PhD       False   
319        Zhu S        Gao B    Male          PhD       False   

         current job     MD  \
0                 PI  False   
1           Academia  False   
2              Admin  False   
3    Facility leader  False   
4                NaN  False   
..               ...    ...   
315     Senior Staff  False   
316     Senior Staff  False   
317               PI  False   
318         Academia  False   
319         Academia  False   

                                           Affiliation  Country Ivy league  \
0                                      Yale University      USA       True   
1                              The University of Texas      USA      False   
2                        Tokyo Metropolitan University    Japan      False   
3                Hungarian Academy of Sciences, Szeged  Hungary      False   
4                Hungarian Academy of Sciences, Szeged  Hungary      False   
..                                                 ...      ...        ...   
315                  niversity of Missouri-Kansas City      USA      False   
316  Institute of Zoology, Chinese Academy of Sciences    China      False   
317  Institute of Zoology, Chinese Academy of Sciences    China      False   
318  Institute of Zoology, Chinese Academy of Sciences    China      False   
319  Institute of Zoology, Chinese Academy of Sciences    China      False   

    first_author_key last_author_key  
0            derre i       agaisse h  
1           seong cs     aguilera rj  
2            tsuda m        aigaki t  
3           markus r          ando i  
4              rus f          ando i  
..               ...             ...  
315             ao j           yu xq  
316           yuan y           zhu s  
317           tian c           zhu s  
318          zhang z           zhu s  
319            gao b           zhu s  

[320 rows x 12 columns]

In [26]:
paper_auth_pairs_LH = paper_auth_pairs[["last author", "last_author_key", "Affiliation", "Country", "Ivy league"]]
paper_auth_pairs_LH = preprocess_utils.deduplicate_by(paper_auth_pairs_LH, "last author").copy()
claims_LH = authors_claims[authors_claims['Authorship'] == 'Leading'].copy()
claims_LH

Name  Historical lab  Historical lab after 1998  Continuity  \
0      Hultmark D            True                        2.0        True   
1      Lemaitre B            True                        1.0        True   
2     Hoffmann JA            True                        2.0        True   
4          Lee WJ           False                        0.0        True   
5           Ip YT           False                        0.0       False   
..            ...             ...                        ...         ...   
153      Nappi AJ            True                        2.0        True   
154      Palli SR           False                        0.0       False   
155     Seroude L           False                        0.0       False   
162  Schneider DS           False                        0.0        True   
163   Schneider D           False                        0.0        True   

     F and L   Sex  Articles  Major claims  Unchallenged  Verified  \
0        2.0  Male      28.0          67.0          12.0      47.0   
1        1.0  Male      21.0          60.0           7.0      47.0   
2        2.0  Male      16.0          45.0           2.0      39.0   
4        0.0  Male      12.0          31.0           8.0      14.0   
5        1.0  Male      11.0          30.0           5.0      21.0   
..       ...   ...       ...           ...           ...       ...   
153      0.0  Male       1.0           1.0           1.0       0.0   
154      0.0  Male       1.0           1.0           0.0       1.0   
155      0.0  Male       1.0           1.0           1.0       0.0   
162      2.0  Male      12.0          31.0          17.0       6.0   
163      2.0  Male       1.0           3.0           0.0       3.0   

     Partially Verified  Mixed  Challenged Start lab        Finish Authorship  
0                   6.0    0.0         2.0         H       Retired    Leading  
1                   5.0    0.0         1.0      1998  still active    Leading  
2                   2.0    0.0         2.0         H       Retired    Leading  
4                   3.0    2.0         4.0      2000  still active    Leading  
5                   3.0    1.0         0.0      1998          2020    Leading  
..                  ...    ...         ...       ...           ...        ...  
153                 0.0    0.0         0.0       NaN           NaN    Leading  
154                 0.0    0.0         0.0       NaN           NaN    Leading  
155                 0.0    0.0         0.0       NaN           NaN    Leading  
162                 1.0    0.0         7.0       NaN           NaN    Leading  
163                 0.0    0.0         0.0       NaN           NaN    Leading  

[157 rows x 16 columns]

In [27]:


# For claims_LH
# drop the row where Name that is NaN 
claims_LH.loc[:, 'last_author_key'] = claims_LH['Name'].str.lower()
claims_LH.loc[:, 'last_author_key'] = (claims_LH['last_author_key']
    .str.normalize('NFKD')
    .str.encode('ascii', errors='ignore')
    .str.decode('utf-8'))
#claims_LH.loc[:, 'last_author_key'] = claims_LH['last_author_key'].str.replace('ando i', 'ando')
claims_LH = clean_author_keys(claims_LH, "last_author_key")


In [28]:
# check for duplicate last_author_key in paper_auth_pairs_LH
paper_auth_pairs_LH[paper_auth_pairs_LH.duplicated('last_author_key', keep=False)]
print(len(paper_auth_pairs_LH[paper_auth_pairs_LH.duplicated('last_author_key', keep=False)]), "duplicates in paper_auth_pairs_LH")
# check for duplicate last_author_key in claims_LH
print(len(claims_LH[claims_LH.duplicated('last_author_key', keep=False)]), "duplicates in claims_LH")

0 duplicates in paper_auth_pairs_LH
0 duplicates in claims_LH


In [29]:
# Perform the outer merge to see what we are missing
all_LH = pd.merge(claims_LH, paper_auth_pairs_LH, on='last_author_key', how='outer')
print(len(claims_LH), len(paper_auth_pairs_LH), len(all_LH))

157 160 161


In [ ]:
unique_pairs = all_LH[["Name", "last author", "last_author_key"]].drop_duplicates().sort_values("last_author_key", ascending=True)
print(" ONLY in Pairs               only in Claims        matching key ")
print("-"*70)
for i in range(0, len(unique_pairs)):
    if pd.isna(unique_pairs.iloc[i]['last author']) or pd.isna(unique_pairs.iloc[i]['Name']):
        print('💥 ', end='')
        print(f"{unique_pairs.iloc[i]['last author']:<20} vs  {unique_pairs.iloc[i]['Name']:<20} --> {unique_pairs.iloc[i]['last_author_key']:<20}")
    #else:
    #    print(f"   {unique_pairs.iloc[i]['last author']:<20} vs  {unique_pairs.iloc[i]['Name']:<20} --> {unique_pairs.iloc[i]['last_author_key']:<20}")

 ONLY in Pairs               only in Claims        matching key 
----------------------------------------------------------------------


NameError: name 'printd' is not defined

In [31]:
all_LH_inner = pd.merge(claims_LH, paper_auth_pairs_LH, on='last_author_key', how='inner').sort_values("last_author_key", ascending=True)
#all_LH_inner.drop(columns=['last_author_key', 'last author', 'Authorship'], inplace=True)
print(len(all_LH_inner))
all_LH_inner.to_csv('preprocessed_data/LH_inner.csv', index=False)
all_LH_inner

156


Name  Historical lab  Historical lab after 1998  Continuity  \
63     Agaisse H           False                        0.0       False   
107  Aguilera RJ           False                        0.0       False   
64      Aigaki T           False                        0.0       False   
6    Anderson KV           False                        0.0       False   
16        Andó I           False                        0.0        True   
..           ...             ...                        ...         ...   
104         Xu T           False                        0.0       False   
146  Yamaguchi M           False                        0.0       False   
105       Yoo MA           False                        0.0       False   
106        Yu XQ           False                        0.0       False   
47         Zhu S           False                        0.0       False   

     F and L     Sex  Articles  Major claims  Unchallenged  Verified  ...  \
63       1.0    Male       1.0           3.0           2.0       1.0  ...   
107      0.0    Male       1.0           2.0           0.0       1.0  ...   
64       0.0    Male       1.0           3.0           2.0       1.0  ...   
6        0.0  Female       8.0          20.0           7.0      11.0  ...   
16       0.0    Male       6.0          12.0           1.0       7.0  ...   
..       ...     ...       ...           ...           ...       ...  ...   
104      0.0    Male       1.0           3.0           0.0       3.0  ...   
146      0.0    Male       1.0           2.0           1.0       1.0  ...   
105      0.0  Female       1.0           3.0           0.0       2.0  ...   
106      0.0    Male       1.0           3.0           1.0       2.0  ...   
47       0.0    Male       4.0           6.0           3.0       2.0  ...   

     Mixed  Challenged  Start lab        Finish Authorship last_author_key  \
63     0.0         0.0        NaN           NaN    Leading       agaisse h   
107    0.0         1.0        NaN           NaN    Leading     aguilera rj   
64     0.0         0.0        NaN           NaN    Leading        aigaki t   
6      0.0         2.0       1997          2010    Leading     anderson kv   
16     0.0         0.0       1999  still active    Leading          ando i   
..     ...         ...        ...           ...        ...             ...   
104    0.0         0.0        NaN           NaN    Leading            xu t   
146    0.0         0.0        NaN           NaN    Leading     yamaguchi m   
105    0.0         0.0        NaN           NaN    Leading          yoo ma   
106    0.0         0.0        NaN           NaN    Leading           yu xq   
47     0.0         0.0       2007          2010    Leading           zhu s   

     last author                                        Affiliation  Country  \
63     Agaisse H                                    Yale University      USA   
107  Aguilera RJ                            The University of Texas      USA   
64      Aigaki T                      Tokyo Metropolitan University    Japan   
6    Anderson KV  Memorial Sloan-Kettering Cancer Center and the...      USA   
16          Ando              Hungarian Academy of Sciences, Szeged  Hungary   
..           ...                                                ...      ...   
104         Xu T                 Yale University School of Medicine      USA   
146  Yamaguchi M             Aichi Cancer Center Research Institute    Japan   
105       Yoo MA                          Pusan National University    Korea   
106        Yu XQ                  niversity of Missouri-Kansas City      USA   
47         Zhu S  Institute of Zoology, Chinese Academy of Sciences    China   

    Ivy league  
63        True  
107      False  
64       False  
6         True  
16       False  
..         ...  
104      False  
146      False  
105      False  
106      False  
47       False  

[156 rows x 21 columns]

In [32]:
for aut in authors[authors["leading_author"]]["author_key"]:
    if len(all_LH[all_LH["last_author_key"] == aut]) != 1:
        print (f"outer: 💥 {aut}, {len(all_LH[all_LH['last_author_key'] == aut])}")
    if len(all_LH_inner[all_LH_inner["last_author_key"] == aut]) != 1:
        print (f"inner: 💥 {aut}, {len(all_LH_inner[all_LH_inner['last_author_key'] == aut])}")


inner: 💥 nappi aj, 0


### B. First authors: TODO

In [33]:
paper_auth_pairs.Affiliation.value_counts()

Affiliation
CNRS-University of Strasbourg                             41
University of Stockholm                                   32
CNRS_Gif-sur-Yvette_Université Paris-Saclay               21
University of Massachusetts Medical School                15
Stanford University Stanford                               7
                                                          ..
The Mount Sinai School of Medicine                         1
Institute of Cancer Research                               1
Massachusetts General Hospital, Harvard Medical School     1
South China Agricultural University                        1
Kanazawa University                                        1
Name: count, Length: 115, dtype: int64

In [34]:
paper_auth_pairs_FH = paper_auth_pairs.drop(columns=['last author']).sort_values("first author", ascending=True)
paper_auth_pairs_FH = preprocess_utils.deduplicate_by(paper_auth_pairs_FH, "first author")

            among ['perrimon n', 'higgins de'].
            among ['Post-doc', 'PhD'].
            among ['Harvard Medical School and Massachusetts General Hospital', 'Massachusetts General Hospital'].
            among ['rahme lg', 'mylonakis e'].
            among ['PhD', 'Post-doc'].
            among ['CEA Grenoble', 'Université de Toulouse'].
            among ['fauvarque mo', 'waltzer l'].
            among ['CNRS Gif-sur-Yvette Université Paris-Saclay', 'CNRS_Gif-sur-Yvette_Université Paris-Saclay'].
            among ['boccard f', 'lemaitre b'].
            among ['CNRS_Gif-sur-Yvette_Université Paris-Saclay', 'CNRS Gif-sur-Yvette Université Paris-Saclay'].
            among ['lemaitre b', 'brey pt'].
            among ['lagueux m', 'hoffmann ja'].
            among ['Massachusetts General Hospital', 'Harvard Medical School,Boston'].
            among ['white k', 'ezekowitz ra'].
            among ['Univeristy of Freiburg', 'Johannes Gutenberg Universität'].
            among [

In [35]:
claims_FH = authors_claims[authors_claims['Authorship'] == 'First'].copy()


claims_FH['first_author_key'] = claims_FH['Name'].str.lower()
claims_FH['first_author_key'] = claims_FH['first_author_key'].str.normalize('NFKD').str.encode('ascii', errors='ignore').str.decode('utf-8')


# check for duplicate first_author_key in paper_auth_pairs_LH
print(len(paper_auth_pairs_FH[paper_auth_pairs_FH.duplicated('first_author_key', keep=False)]), "duplicates in paper_auth_pairs_LH")
# check for duplicate first_author_key in claims_LH
print(len(claims_FH[claims_FH.duplicated('first_author_key', keep=False)]), "duplicates in claims_LH")


all_FH = pd.merge(claims_FH, paper_auth_pairs_FH, on='first_author_key', how='outer')
print(len(claims_FH), len(paper_auth_pairs_FH), len(all_FH))

print(" ONLY in Pairs               only in Claims        matching key ")
print("-"*70)
unique_pairs = all_FH[["Name", "first author", "first_author_key"]].drop_duplicates().sort_values("first_author_key", ascending=True)
for i in range(0, len(unique_pairs)):
    if pd.isna(unique_pairs.iloc[i]['first author']) or pd.isna(unique_pairs.iloc[i]['Name']):
        print('💥 ', end='')
        print(f" {unique_pairs.iloc[i]['first author']:<20}  {unique_pairs.iloc[i]['Name']:<20} --> {unique_pairs.iloc[i]['first_author_key']:<20}")

0 duplicates in paper_auth_pairs_LH
4 duplicates in claims_LH
292 291 294
 ONLY in Pairs               only in Claims        matching key 
----------------------------------------------------------------------
💥  nan                   Hedengren-Olcott M   --> hedengren-olcott m  
💥  RIZKI MT              nan                  --> rizki mt            
💥  Schneider D           nan                  --> schneider d         


In [52]:
def safe_strip(x):
    """
    Safely strip whitespace from strings in a DataFrame column.
    """
    if x.dtype == "object":
        return x.apply(lambda val: val.strip() if isinstance(val, str) else val)
    return x

all_FH = all_FH.apply(safe_strip)

In [53]:
all_FH_inner = pd.merge(claims_FH, paper_auth_pairs_FH, on='first_author_key', how='inner').sort_values("first_author_key", ascending=True)
all_FH_inner.drop(columns=['Continuity', 'Historical lab after 1998', 'Historical lab', "Sex_y"], inplace=True)
# rename Sex_x to Sex
all_FH_inner = all_FH_inner.rename(columns={'Sex_x': 'Sex'})
print(len(all_FH_inner))
all_FH_inner.to_csv('preprocessed_data/FH_inner.csv', index=False)
all_FH_inner

291


Name  F and L     Sex  Articles  Major claims  Unchallenged  \
80    Abdelsadik A      NaN    Male       1.0           3.0           3.0   
19       Agaisse H      NaN    Male       2.0           6.0           2.0   
196     Aggarwal K      NaN  Female       1.0           2.0           0.0   
81            Ao J      NaN    Male       1.0           3.0           1.0   
20   Apidianakis Y      NaN    Male       3.0           6.0           4.0   
..             ...      ...     ...       ...           ...           ...   
194  Zettervall CJ      NaN    Male       1.0           3.0           0.0   
273        Zhang Z      NaN  Female       2.0           2.0           1.0   
195        Zhao HW      NaN  Female       1.0           3.0           2.0   
290         Zhou Z      NaN    Male       1.0           1.0           0.0   
42       Zhuang ZH      NaN    Male       2.0           6.0           1.0   

     Verified  Partially Verified  Mixed  Challenged  ... first_author_key  \
80        0.0                 0.0    0.0         0.0  ...     abdelsadik a   
19        4.0                 0.0    0.0         0.0  ...        agaisse h   
196       2.0                 0.0    0.0         0.0  ...       aggarwal k   
81        2.0                 0.0    0.0         0.0  ...             ao j   
20        2.0                 0.0    0.0         0.0  ...    apidianakis y   
..        ...                 ...    ...         ...  ...              ...   
194       3.0                 0.0    0.0         0.0  ...    zettervall cj   
273       1.0                 0.0    0.0         0.0  ...          zhang z   
195       1.0                 0.0    0.0         0.0  ...          zhao hw   
290       1.0                 0.0    0.0         0.0  ...           zhou z   
42        4.0                 1.0    0.0         0.0  ...        zhuang zh   

      first author  PhD Post-doc Become a Pi   current job     MD  \
80    Abdelsadik A  Senior Staff        True            PI  False   
19       Agaisse H      Post-doc        True            PI  False   
196     Aggarwal K           PhD       False      Industry  False   
81            Ao J           PhD         NaN  Senior Staff  False   
20   Apidianakis Y      Post-doc        True            PI  False   
..             ...           ...         ...           ...    ...   
194  Zettervall CJ           PhD       False      Industry  False   
273        Zhang Z           PhD       False      Academia  False   
195        Zhao HW           PhD       False           NaN  False   
290         Zhou Z           PhD         NaN           NaN  False   
42       Zhuang ZH           PhD       False           NaN  False   

                                           Affiliation  Country Ivy league  \
80                           Forschungszentrum Borstel  Germany      False   
19                              Harvard Medical School      USA       True   
196         University of Massachusetts Medical School      USA      False   
81                   niversity of Missouri-Kansas City      USA      False   
20   Harvard Medical School and Massachusetts Gener...      USA       True   
..                                                 ...      ...        ...   
194                            University of Stockholm   Sweden      False   
273  Institute of Zoology, Chinese Academy of Sciences    China      False   
195                 University of California San Diego      USA       True   
290                              University of Houston      USA      False   
42                       Shanghai Jiao Tong University    China      False   

    last_author_key  
80         roeder t  
19       perrimon n  
196     silverman n  
81            yu xq  
20         rahme lg  
..              ...  
194      hultmark d  
273           zhu s  
195       haddad gg  
290     kimbrell da  
42            ge bx  

[291 rows x 23 columns]

In [54]:
for aut in authors[authors["first_author"]]["author_key"]:
    #if len(all_FH[all_LH["first_author_key"] == aut]) != 1:
    #    print (f"outer: 💥 {aut}, {len(all_FH[all_FH['author_key'] == aut])}")
    if len(all_FH_inner[all_FH_inner["first_author_key"] == aut]) != 1:
        print (f"inner: 💥 {aut}, {len(all_FH_inner[all_FH_inner['first_author_key'] == aut])}")

inner: 💥 hedengren-olcott m, 0
inner: 💥 hedengren m, 2
inner: 💥 hedengren m, 2
inner: 💥 kim ys, 2
inner: 💥 kim ys, 2
inner: 💥 kim ys, 2


In [49]:
all_FH_inner.columns

Index(['Name', 'F and L', 'Sex', 'Articles', 'Major claims', 'Unchallenged',
       'Verified', 'Partially Verified', 'Mixed', 'Challenged', 'Start lab',
       'Finish', 'Authorship', 'first_author_key', 'first author',
       'PhD Post-doc', 'Become a Pi', 'current job', 'MD', 'Affiliation',
       'Country', 'Ivy league', 'last_author_key'],
      dtype='object')

In [55]:
for col in all_FH_inner.columns:
    print(70*"-")
    print(all_FH_inner[col].value_counts())

----------------------------------------------------------------------
Name
Hedengren M     2
Kim YS          2
Abdelsadik A    1
Paquette N      1
Radyuk SN       1
               ..
Honti V         1
Helenius IT     1
Han ZS          1
Han SH          1
Zhuang ZH       1
Name: count, Length: 289, dtype: int64
----------------------------------------------------------------------
Series([], Name: count, dtype: int64)
----------------------------------------------------------------------
Sex
Male      168
Female    123
Name: count, dtype: int64
----------------------------------------------------------------------
Articles
1.0    213
2.0     55
3.0     17
4.0      5
5.0      1
Name: count, dtype: int64
----------------------------------------------------------------------
Major claims
3.0     116
2.0      79
6.0      24
5.0      24
1.0      17
4.0      12
9.0       6
8.0       5
7.0       4
10.0      2
12.0      1
14.0      1
Name: count, dtype: int64
----------------------------------

## match the pairs in DV

In [38]:
paper_auth_pairs

last author first author     Sex PhD Post-doc Become a Pi  \
0      Agaisse H        Derré  Female     Post-doc        True   
1    Aguilera RJ     Seong CS    Male          PhD       False   
2       Aigaki T      Tsuda M    Male          PhD        True   
3           Ando       Markus    Male          PhD       False   
4           Ando        Rus F  Female          PhD       False   
..           ...          ...     ...          ...         ...   
315        Yu XQ         Ao J     NaN          PhD         NaN   
316        Zhu S       Yuan Y    Male          PhD       False   
317        Zhu S       Tian C  Female          PhD        True   
318        Zhu S      Zhang Z  Female          PhD       False   
319        Zhu S        Gao B    Male          PhD       False   

         current job     MD  \
0                 PI  False   
1           Academia  False   
2              Admin  False   
3    Facility leader  False   
4                NaN  False   
..               ...    ...   
315     Senior Staff  False   
316     Senior Staff  False   
317               PI  False   
318         Academia  False   
319         Academia  False   

                                           Affiliation  Country Ivy league  \
0                                      Yale University      USA       True   
1                              The University of Texas      USA      False   
2                        Tokyo Metropolitan University    Japan      False   
3                Hungarian Academy of Sciences, Szeged  Hungary      False   
4                Hungarian Academy of Sciences, Szeged  Hungary      False   
..                                                 ...      ...        ...   
315                  niversity of Missouri-Kansas City      USA      False   
316  Institute of Zoology, Chinese Academy of Sciences    China      False   
317  Institute of Zoology, Chinese Academy of Sciences    China      False   
318  Institute of Zoology, Chinese Academy of Sciences    China      False   
319  Institute of Zoology, Chinese Academy of Sciences    China      False   

    first_author_key last_author_key  
0            derre i       agaisse h  
1           seong cs     aguilera rj  
2            tsuda m        aigaki t  
3           markus r          ando i  
4              rus f          ando i  
..               ...             ...  
315             ao j           yu xq  
316           yuan y           zhu s  
317           tian c           zhu s  
318          zhang z           zhu s  
319            gao b           zhu s  

[320 rows x 12 columns]

In [39]:
authors

author_id  article_id               name     sex  leading_author  \
0            2        2053             Brun S    Male           False   
1            3        2053         Lemaitre B    Male            True   
2            1        1650         Charroux B    Male           False   
3           97        2149        Söderhäll K    Male            True   
4            5        2278        Silverman N    Male           False   
..         ...         ...                ...     ...             ...   
792        797        2067         Cerenius L    Male            True   
793        798        2122         Higgins DE    Male            True   
794        799        2184  Mengin-Lecreulx D    Male            True   
795        800        2130           Moore KJ  Female            True   
796        700        1990           Matova N  Female           False   

     first_author career_stage expertise_level         author_key  
0            True     Post-doc     Experienced             brun s  
1           False    Junior PI     Experienced         lemaitre b  
2            True     Post-doc     Experienced         charroux b  
3           False    Senior PI     Experienced        soderhall k  
4            True     Post-doc        Newcomer        silverman n  
..            ...          ...             ...                ...  
792         False    Senior PI     Experienced         cerenius l  
793         False    Senior PI        Newcomer         higgins de  
794         False    Senior PI     Experienced  mengin-lecreulx d  
795         False    Junior PI        Newcomer           moore kj  
796          True     Post-doc        Newcomer           matova n  

[797 rows x 9 columns]

In [40]:

paper_auth_pairs["article_id"] = pd.NA
paper_auth_pairs["article_id"] = paper_auth_pairs["article_id"].astype("Int64")

for art in authors["article_id"].unique():
    these_authors = authors[authors["article_id"] == art]
    these_first_authors = these_authors[these_authors["first_author"] == True]["author_key"].values
    these_leading_authors = these_authors[these_authors["leading_author"] == True]["author_key"].values
    if (len(these_first_authors) == 1 and len(these_leading_authors) == 1):
        indexes = paper_auth_pairs[(paper_auth_pairs["last_author_key"] == these_leading_authors[0]) & 
                                (paper_auth_pairs["first_author_key"] == these_first_authors[0])].index.to_list()
        
        if len(indexes) == 1:
            paper_auth_pairs.loc[indexes[0], "article_id"] = art
        elif len(indexes) > 1:
            first_na = pd.isna(paper_auth_pairs.loc[indexes, "article_id"]).idxmax() #idmax returns the index of the first occurrence of the maximum value
            print(these_authors["author_key"].values, pd.isna(paper_auth_pairs.loc[indexes, "article_id"]).values, first_na)
            paper_auth_pairs.loc[first_na, "article_id"] = art
        else:
            print(f"NO MATCH 🛑 {art}, {len(indexes)}, {these_first_authors[0]:<20} -- {these_leading_authors[0]:<20} ... {len(these_authors)}")
    else:
        print(f"SKIPPING {art} because only one={len(these_authors)} author: {these_authors['author_key'].values}")

['georgel p' 'hoffmann ja'] [ True  True] 96
NO MATCH 🛑 2239, 0, hedengren-olcott m   -- taylor bj            ... 2
['georgel p' 'hoffmann ja'] [False  True] 101
NO MATCH 🛑 1947, 0, schneider ds         -- schneider ds         ... 1
NO MATCH 🛑 2730, 0, schneider ds         -- schneider ds         ... 2
NO MATCH 🛑 3447, 0, rizki tm             -- rizki tm             ... 1
NO MATCH 🛑 3448, 0, rizki tm             -- rizki tm             ... 1
NO MATCH 🛑 3292, 0, rizki tm             -- rizki tm             ... 1
NO MATCH 🛑 3273, 0, rizki tm             -- rizki tm             ... 2
['georgel p' 'hoffmann ja'] [False False] 96
NO MATCH 🛑 2764, 0, imler jl             -- imler jl             ... 1
NO MATCH 🛑 3446, 0, nappi aj             -- nappi aj             ... 1


### There is an error above, : 3 hoffman georgel, whereas in the excel there are two...

 ## 2. Process journal articles from the database

In [41]:
paper_auth_pairs[paper_auth_pairs["article_id"].isna()]

last author  first author     Sex PhD Post-doc Become a Pi  \
16       Bellotti RA      Rizki TM    Male           PI        True   
106      Hoffmann JA      Imler JL    Male           PI        True   
240         Rizki TM      RIZKI MT    Male           PI        True   
271    Shahabuddin M   Schneider D    Male           Pi        True   
272  Shirasu-Hiza MM  Schneider DS    Male           Pi        True   
278       Silvers MJ      Nappi AJ    Male           PI        True   
290        Taylor BJ   Hedengren M  Female          PhD       False   

    current job     MD                    Affiliation Country Ivy league  \
16     Deceased  False                      Ann arbor     USA      False   
106          PI  False  CNRS-University of Strasbourg  France      False   
240     Retired  False                      ann arbor     NaN      False   
271          PI   True            Whitehead Institute     USA      False   
272          PI   True   Stanford University Stanford     USA       True   
278          PI  False          Univeristy of Chicago     USA      False   
290         NaN  False        Oregon State University     USA      False   

    first_author_key  last_author_key  article_id  
16          rizki tm      bellotti ra        <NA>  
106         imler jl      hoffmann ja        <NA>  
240         rizki mt         rizki tm        <NA>  
271      schneider d    shahabuddin m        <NA>  
272     schneider ds  shirasu-hiza mm        <NA>  
278         nappi aj       silvers mj        <NA>  
290      hedengren m        taylor bj        <NA>

In [42]:
paper_auth_pairs.loc[290, "article_id"] = 2239

In [43]:
paper_auth_pairs[paper_auth_pairs["article_id"].isna()]

last author  first author   Sex PhD Post-doc Become a Pi current job  \
16       Bellotti RA      Rizki TM  Male           PI        True    Deceased   
106      Hoffmann JA      Imler JL  Male           PI        True          PI   
240         Rizki TM      RIZKI MT  Male           PI        True     Retired   
271    Shahabuddin M   Schneider D  Male           Pi        True          PI   
272  Shirasu-Hiza MM  Schneider DS  Male           Pi        True          PI   
278       Silvers MJ      Nappi AJ  Male           PI        True          PI   

        MD                    Affiliation Country Ivy league first_author_key  \
16   False                      Ann arbor     USA      False         rizki tm   
106  False  CNRS-University of Strasbourg  France      False         imler jl   
240  False                      ann arbor     NaN      False         rizki mt   
271   True            Whitehead Institute     USA      False      schneider d   
272   True   Stanford University Stanford     USA       True     schneider ds   
278  False          Univeristy of Chicago     USA      False         nappi aj   

     last_author_key  article_id  
16       bellotti ra        <NA>  
106      hoffmann ja        <NA>  
240         rizki tm        <NA>  
271    shahabuddin m        <NA>  
272  shirasu-hiza mm        <NA>  
278       silvers mj        <NA>

In [44]:
len(paper_auth_pairs.article_id.unique())

315

## Summary
So how to analyze ? The key works well to merge at least, so let's have it. I think we should not do by authors because the aurhors are fleeting. Whether it's first or not. Let's do it by claim when it is needed. I think it's much more correct.